pkg/reloader/reloader.go

```diff
@@ -0,0 +1,212 @@
+	lastRulesHash    []byte
+}
+
+// New creates a new reloader that watches the given input file and copies
+// it to the output file path.
+// It optionally substitutes environment variables on copying.
+func New(logger log.Logger, promURL *url.URL, cfgFile, cfgEnvsubstFile, ruleDir string) *Reloader {
+	return &Reloader{
+		logger:           logger,
+		promURL:          promURL,
+		cfgFilename:      cfgFile,
+		cfgSubstFilename: cfgEnvsubstFile,
+	}
+}
+
+// Watch starts to watch the config file and processes it until the context
+// gets canceled.
+func (r *Reloader) Watch(ctx context.Context) error {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return errors.Wrap(err, "create watcher")
+	}
+	defer w.Close()
+
+	cfgDir := filepath.Dir(r.cfgFilename)
+
+	if err := w.Add(cfgDir); err != nil {
+		return errors.Wrap(err, "add file watch")
+	}
+	if _, err := r.applyConfig(); err != nil {
+		return errors.Wrap(err, "initial apply")
+	}
+	if err := r.triggerReload(ctx); err != nil {
+		level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+	}
+
+	tick := time.NewTicker(3 * time.Minute)
+	defer tick.Stop()
+
+	for {
+		select {
+		case event := <-w.Events:
+			level.Debug(r.logger).Log("msg", "received watch event", "op", event.Op, "name", event.Name)
+
+			if event.Name != r.cfgFilename {
+				continue
+			}
+			changes, err := r.applyConfig()
+			if err != nil {
+				level.Error(r.logger).Log("msg", "apply failed", "err", err)
+			}
+			if !changes {
+				continue
+			}
+			if err := r.triggerReload(ctx); err != nil {
+				level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+			}
+
+		case err := <-w.Errors:
+			level.Error(r.logger).Log("msg", "watch error", "err", err)
+
+		case <-tick.C:
+			changes, err := r.refreshRules()
+			if err != nil {
+				level.Error(r.logger).Log("msg", "refreshing rules failed", "err", err)
+			}
+			if !changes {
+				continue
+			}
+			if err := r.triggerReload(ctx); err != nil {
+				level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+			}
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func hashFile(h hash.Hash, fn string) error {
+	f, err := os.Open(fn)
+	if err != nil {
+		return err
+	}
+	h.Write([]byte{'\xff'})
+	h.Write([]byte(fn))
+	h.Write([]byte{'\xff'})
+
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reloader) applyConfig() (ok bool, err error) {
+	if r.cfgFilename == "" {
+		return false, nil
+	}
+	h := sha256.New()
+
+	if err := hashFile(h, r.cfgFilename); err != nil {
+		return false, errors.Wrap(err, "hash file")
+	}
+	hb := h.Sum(nil)
+
+	if bytes.Equal(r.lastCfgHash, hb) {
+		return false, nil
+	}
+	r.lastCfgHash = hb
+
+	if r.cfgSubstFilename == "" {
+		return true, nil
+	}
+
+	b, err := envsubst.ReadFileRestricted(r.cfgFilename, true, true)
+	if err != nil {
+		return false, errors.Wrap(err, "read file")
+	}
+	if err := ioutil.WriteFile(r.cfgSubstFilename, b, 0666); err != nil {
```

Comment:
Why we have this file `cfgSubstFilename`? Why we don't use just `cfgFilename`?

Reply:
Seems kinda sketchy to overwrite a configmap-mounted file. I'm not confident about at which point it just might get overwritten again – which might then race with the reload we triggered.

Also we'd retrigger our own file watch again and would have to detect that properly.

Reply:
OK, can we at least name it differently? `cfgInpuFIlename` and `cfgOutputFilename` files or something like that?

Reply:
We only substitute if there's a file name set for the substitution. I don't really wanna do the copy step (and required config overhead) if people don't need it. Having is just work as enabling it and providing the target file at once seemed simpler than adding another boolean flag, that only is valid if the output filename is set and the output filename not being sensical if the flag is false...

Reply:
I clarified the docs of the main type a bit.